### PR TITLE
Embed data files

### PR DESCRIPTION
**Pull request by Avi-D-coder**
closes:#699

Embeds data files so the hlint executable is self contained.
This allows for easy packaging with stack and cabal.
And enables hlint to be used as a library without shipping data files.

From https://github.com/ndmitchell/hlint/issues/620#issuecomment-570777653
> ndmitchell would you reconsider the TH approach?
> 
> It's not just nicer for using `hlint` as a library it's nicer when using the executable too. I'm on a 256G laptop and I delete `~/.stack` when ever it gets past 10G. I install hlint using `stack install`, so when I delete `~/.stack` `hlint` has to be rebuilt, just to get 3 data files.
> 
> Here's a [commit](https://github.com/Avi-D-coder/hlint/commit/b50d9cf08e409d4ca20abf37fb541baebff60391) using `file-embed`. The TH is relegated to one file. The executable actually gets slightly smaller `48117184` to `48020968` (GHC 8.8).

### CODE REVIEW

**Comment by Avi-D-coder** [comment]
`TemplateHaskell` is confined to `src/EmbedData.hs`. If it causes a problem on a particular platform, an alternative module could be conditionally compiled with the old `readFile $ cmdDataDir </> "default.yaml"` method.

File: src/CmdLine.hs [hunk 1]
```diff
@@ -217,12 +217,11 @@ mode = cmdArgsMode $ modes
 --   We want more important hints to go last, since they override
 cmdHintFiles :: Cmd -> IO [FilePath]
 cmdHintFiles cmd = do
-    let explicit1 = [cmdDataDir cmd </> "hlint.yaml" | null $ cmdWithHints cmd]
```
Comment (ndmitchell) [inline src/CmdLine.hs]: If `cmdWithHints` is specified then we don't use `cmdDataDir` for the hints. It seems in your patch that changes?

File: src/Config/Read.hs [hunk 2]
```diff
@@ -6,12 +6,14 @@ import Config.Haskell
...
 
 readFilesConfig :: [(FilePath, Maybe String)] -> IO [Setting]
 readFilesConfig files = do
-        yaml <- mapM (uncurry readFileConfigYaml) yaml
+        yaml <- mapM (uncurry readFileConfigYaml) yaml'
         haskell <- mapM (uncurry readFileConfigHaskell) haskell
         return $ concat haskell ++ settingsFromConfigYaml yaml
     where
         (yaml, haskell) = partition (\(x,_) -> lower (takeExtension x) `elem` [".yml",".yaml"]) files
+        yaml' = if any ((takeFileName (fst hlintYaml) ==) . takeFileName . fst) yaml then yaml else hlintYaml : yaml
```
Comment (ndmitchell) [inline src/Config/Read.hs]: I don't get what this line is doing

File: hlint.cabal [hunk 3]
```diff
@@ -54,6 +54,8 @@ library
         base == 4.*, process, filepath, directory, containers,
         unordered-containers, vector, text, bytestring,
         transformers,
+        file-embed,
+        utf8-string,
```
Comment (ndmitchell) [inline hlint.cabal]: Where is the utf8-string dependency coming from?

Comment (Avi-D-coder) [reply hlint.cabal]: embed stores a `ByteString` of the UTF-8 data files, so we have to convert it back.

File: src/CmdLine.hs (hunk 1 above)
Comment (Avi-D-coder) [reply src/CmdLine.hs]: My bad, should be fixed now

**Comment by ndmitchell** [comment]
That's way nicer. Thanks a lot!

**Comment by Avi-D-coder** [comment]
@ndmitchell Could you push a release to hackage so hie can use this?

**Comment by Avi-D-coder** [comment]
@ndmitchell Could you push a release to hackage so hie can use this?

**Comment by ndmitchell** [comment]
Sorry @Avi-D-coder - I've been on holiday. Should have this done by this evening.

**Comment by Avi-D-coder** [comment]
Great thanks

**Pull request merged**